network/benchmark/src/lib.rs

```diff
@@ -295,6 +279,89 @@ pub async fn run_netbench_service(
     }
 }
 
+async fn connection_listener(
+    node_config: NodeConfig,
+    network_client: NetworkClient<NetbenchMessage>,
+    time_service: TimeService,
+    shared: Arc<RwLock<NetbenchSharedState>>,
+    handle: Handle,
+) {
+    let config = node_config.netbench.unwrap();
+    let peers_and_metadata = network_client.get_peers_and_metadata();
+    if let Ok(peers) = peers_and_metadata.get_connected_peers_and_metadata() {
+        info!(
+            "netbench connection_listener got {} initial peers",
+            peers.len()
+        );
+        for peer_network_id in peers.keys() {
+            info!(
+                "netbench connection_listener new for initial {:?}",
+                peer_network_id
+            );
+            if config.enable_direct_send_testing {
+                handle.spawn(direct_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+            if config.enable_rpc_testing {
+                handle.spawn(rpc_sender(
+                    node_config.clone(),
+                    network_client.clone(),
+                    time_service.clone(),
+                    peer_network_id.network_id(),
+                    peer_network_id.peer_id(),
+                    shared.clone(),
+                ));
+            }
+        }
+    }
+    let mut connection_notifications = peers_and_metadata.subscribe();
+    loop {
+        match connection_notifications.recv().await {
+            None => {
+                info!("netbench connection_listener exit");
+                return;
+            },
+            Some(note) => match note {
+                ConnectionNotification::NewPeer(meta, netcontext) => {
```

Comment:
IIUC, this is the only place that reads the network context, right? If so, I think it could be removed from the `NewPeer` and `LostPeer` events. The peer ID should be in `metadata` already, and it looks like the network id is only used for netbench logging?

Reply:
We could probably further streamline `ConnectionNotification` towards what it is actually used for right now, but maybe push that back for a round-2?

Reply:
Hmm.. I'd think doing it here feels cleaner. The PR isn't that big. But, your call.

Reply:
I'm just afraid of scope creep leading to unbounded delay. Better to get _something_ in.

Reply:
After cleanup and some secondary unused things getting deleted, the next PR is up to 14 files, +74-168, and I think it deserves to be a next PR

Reply:
the follow on diff if you want to peek ahead
https://github.com/brianolson/aptos-core/pull/2

Reply:
e.g. this PR is already the expanded form of https://github.com/aptos-labs/aptos-core/pull/12617 , it has to stop somewhere and merge and make a next PR, otherwise we get all the way to network2 branch and it never merges.

Reply:
Ok. Finishing the job and doing one more round of simplify&cleanup moves this from a +434-254 change to a +373-292 change, win! Let's do that! Going to merge the further stuff which had been a follow-on PR into this PR.